core/src/main/scala/org/apache/spark/JobArtifactSet.scala

```diff
@@ -53,72 +56,70 @@ class JobArtifactSet(
           this.files.toSeq == that.files.toSeq && this.archives.toSeq == that.archives.toSeq
     }
   }
-
 }
 
-object JobArtifactSet {
 
-  private[this] val current = new ThreadLocal[Option[JobArtifactSet]] {
-    override def initialValue(): Option[JobArtifactSet] = None
-  }
+private[spark] object JobArtifactSet {
+  // For testing.
+  val emptyJobArtifactSet: JobArtifactSet = new JobArtifactSet(
+    None, None, Map.empty, Map.empty, Map.empty)
+  // For testing.
+  def defaultJobArtifactSet: JobArtifactSet = SparkContext.getActive.map(
+    getActiveOrDefault).getOrElse(emptyJobArtifactSet)
 
-  /**
-   * When Spark Connect isn't used, we default back to the shared resources.
-   * @param sc The active [[SparkContext]]
-   * @return A [[JobArtifactSet]] containing a copy of the jars/files/archives from the underlying
-   *         [[SparkContext]] `sc`.
-   */
-  def apply(sc: SparkContext): JobArtifactSet = {
-    new JobArtifactSet(
-      uuid = None,
-      replClassDirUri = sc.conf.getOption("spark.repl.class.uri"),
-      jars = sc.addedJars.toMap,
-      files = sc.addedFiles.toMap,
-      archives = sc.addedArchives.toMap)
+  private[this] val currentSessionUUID = new ThreadLocal[Option[String]] {
+    override def initialValue(): Option[String] = None
   }
 
-  private lazy val emptyJobArtifactSet = new JobArtifactSet(
-    None,
-    None,
-    Map.empty,
-    Map.empty,
-    Map.empty)
-
-  /**
-   * Empty artifact set for use in tests.
-   */
-  private[spark] def apply(): JobArtifactSet = emptyJobArtifactSet
-
-  /**
-   * Used for testing. Returns artifacts from [[SparkContext]] if one exists or otherwise, an
-   * empty set.
-   */
-  private[spark] def defaultArtifactSet(): JobArtifactSet = {
-    SparkContext.getActive.map(sc => JobArtifactSet(sc)).getOrElse(JobArtifactSet())
+  private[this] val currentReplClassDirURI = new ThreadLocal[Option[String]] {
+    override def initialValue(): Option[String] = None
   }
 
+  def getCurrentSessionUUID: Option[String] = currentSessionUUID.get()
+
   /**
-   * Execute a block of code with the currently active [[JobArtifactSet]].
-   * @param active
-   * @param block
-   * @tparam T
+   * Set the Spark Connect specific information in the active client to the underlying
+   * [[JobArtifactSet]].
+   *
+   * @param uuid UUID to use in the current context of jab artifact set. Usually this is from
+   *             a Spark Connect client.
+   * @param classDirUri The URI for the directory that stores REPL classes.
+   * @return the result from the function applied with [[JobArtifactSet]] specific to
+   *         the active client.
    */
-  def withActive[T](active: JobArtifactSet)(block: => T): T = {
-    val old = current.get()
-    current.set(Option(active))
+  def withActive[T](uuid: String, classDirUri: Option[String])(block: => T): T = {
```

Comment:
IMO this feels a bit weird to use `withActive` this way because we are passing in a few of the arguments to create the `JobArtifactSet` but we don't create one and instead of rely on the eventual call to `getActiveOrDefault` to create the final instance.

Why not have a helper method to create an instance and use the instance itself in `withActive`? 
If we want to limit what the contents of `jars/files/archives` vals are, we could make the default constructor private and force creation through `apply` or helper methods.